3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -72,233 +74,288 @@ object EnrichmentManager {
+    ///////////////////////////
 
     // Let's start populating the CanonicalOutput
     // with the fields which cannot error
     val event = setupEnrichedEvent(raw, etlTstamp, processor)
 
-    // 2. Enrichments which can fail
+    // Map/decode the fields of the input event to the enriched event
+    val sourceMap: SourceMap = raw.parameters
+    val firstPassTransform = event.transform(sourceMap, firstPassTransformMap)
+    val secondPassTransform = event.transform(sourceMap, secondPassTransformMap)
 
-    // 2a. Failable enrichments which don't need the payload
+    val transform = EitherT(
+      for {
+        first <- Monad[F].pure(firstPassTransform)
+        second <- Monad[F].pure(secondPassTransform)
+      } yield (first, second)
+        .mapN { (_, _) =>
+          ()
+        }
+        .leftMap { enrichmentFailures =>
+          buildEnrichmentFailuresBadRow(
+            enrichmentFailures,
+            EnrichedEvent.toPartiallyEnrichedEvent(event),
+            RawEvent.toRawEvent(raw),
+            processor
+          )
+        }
+        .toEither
+    )
 
-    // Validate that the collectorTstamp exists and is Redshift-compatible
-    val collectorTstamp: Either[FailureDetails.EnrichmentStageIssue, Unit] =
-      setCollectorTstamp(event, raw.context.timestamp)
+    (for {
+      _ <- transform
 
-    // Attempt to decode the useragent
-    // May be updated later if we have a `ua` parameter
-    val useragent: Either[FailureDetails.EnrichmentStageIssue, Unit] =
-      setUseragent(event, raw.context.useragent, raw.source.encoding)
+      //////////////////////////////////////////////////////////////
+      // 2. Check if input contexts and unstructured event are valid
+      //////////////////////////////////////////////////////////////
 
-    // 2b. Failable enrichments using the payload
-    val sourceMap: SourceMap = raw.parameters
-    val firstPassTransform = event.transform(sourceMap, firstPassTransformMap)
-    val secondPassTransform = event.transform(sourceMap, secondPassTransformMap)
+      inputSDJs <- EitherT(
+        for {
+          contexts <- IgluUtils.extractAndValidateInputContexts(event, client)
+          unstruct <- IgluUtils.extractAndValidateUnstructEvent(event, client).map(_.toValidatedNel)
+        } yield (contexts, unstruct)
+          .mapN { (c, ue) =>
+            (c, ue)
+          }
+          .leftMap { schemaViolations =>
+            buildSchemaViolationsBadRow(
+              schemaViolations,
+              EnrichedEvent.toPartiallyEnrichedEvent(event),
+              RawEvent.toRawEvent(raw),
+              processor
+            )
+          }
+          .toEither
+      )
 
-    // The load fails if the collector version is not set
-    val collectorVersionSet: Either[FailureDetails.EnrichmentStageIssue, Unit] =
-      getCollectorVersionSet(event)
+      ////////////////////////////////////////////////////////
+      // 3. Run enrichments and check that they are successful
```

Comment:
Can we move the whole 3 into a separate function, something like:

```scala
def runEnrichments(event: EnrichedEvent, inputContexts: List[SelfDescribingData[Json]]): EitherT[F, BadRow, List[SelfDescribingData[Json]]] = ???
```

And then this whole 300LoC function would look like:

```scala
for {
  _                              <- transform
  (inputContexts, unstructEvent) <- IgluUtils.extractTrackedJsons(event, client)
  derivedContexts                <- runEnrichments(event, inputContexts)
  _                              <- IgluUtils.validateEnrichmentsContexts[F](client, derivedContexts)
  _                              <- anonymize(event)
} yield event
```

Which is basically an essence of the enrichment.